### ISSUE TITLE
Clarification on CAN Image Conversion and Reproducibility of Results

### ISSUE THREAD

**Issue by alishan2040** [evenchen6/CAN_GAN_Anomaly#3]
Hi @evenchen6, thank you for releasing the code.

I am trying to reproduce the results reported in your paper using the CAN-bus dataset and have a few questions regarding the implementation:

**CAN Image Conversion:**
I followed the documentation and your comment suggesting that the conversion of raw CAN data to images refers to the method described in the paper figure. I implemented a script that converts raw CAN IDs into 48x48 image format for training. However, I am unsure if this aligns exactly with your approach. Could you clarify the following?
- Is there any additional preprocessing or normalization step beyond reshaping into 48x48?

**Reproducibility:**
Using the same training/validation/test distribution as in the paper and the best model strategy you provided (ACGAN-AC-RFC-Th), I obtained the following results on the test set:
- Macro F1: ~0.5–0.6
- Unknown F1 (DoS): ~0.2–0.3

These are significantly lower than the reported >0.99 scores in your paper. I’m wondering if this discrepancy could be due to a mismatch in the image generation process or if there are other critical implementation details missing from the public release?

**Applicability to Other Datasets:**
I’m also exploring the application of your method to other intrusion detection datasets, such as CICIDS2017. If a dataset has 30 features, would it be acceptable to construct a 6x6 image (with zero-padding) for model input? Or do you recommend a different approach for non-CAN datasets?

Any clarification or pointers would be greatly appreciated.

**Comment by evenchen6** [comment]
Hi @alishan2040,

Thank you very much for your detailed questions and effort in reproducing our work. Please see our clarifications below:

## **1. CAN Image Conversion**

Regarding your first question on CAN image conversion, according to Section 3.1 of the paper ("Dataset Construction"), each CAN image is constructed by grouping a consecutive sequence of N = 48 CAN messages into a 48 × 48 image, as illustrated in Figure 4(b). There are two main sliding window approaches for generating these images:

- **Sliding by one position** creates overlapping images,
- **Sliding by 48 positions** creates non-overlapping images.

In our implementation and as described in the paper, we use the approach where the window slides by 48 positions – that is, each CAN image is built from a distinct set of 48 consecutive CAN messages, with no overlap between images.

Additionally, when converting to CAN images, if any one of the 48 CAN messages in a window contains an attack frame, we label the entire image as a known attack (KA) or unknown attack (UA), depending on the type. Otherwise, the image is labeled as normal.

## **2. Reproducibility of Results**

To better assist you, I would like to clarify your experimental setup—are you evaluating using our provided pre-trained weights, or did you retrain the model from scratch using your own constructed CAN images?

### **Scenario 1: Using Our Provided Weights**

If you are using our publicly released model weights, I suspect the performance degradation is primarily caused by inconsistencies in the CAN image conversion process. In deep learning practice, model performance relies heavily on the consistency of data distribution. If the input data construction method (such as sliding window stride, label definition, normalization method, etc.) differs from what was used during training, it may introduce distribution shift, thereby affecting the model's evaluation results.

To verify this, I re-ran the evaluation using the provided weights on our standard validation set, and the results are completely consistent with those reported in the paper. I will upload the validation logs shortly, which you can load and visualize using VisualDL to examine the detailed evaluation metrics.

**Recommendations:**
- Please carefully verify that your CAN image construction pipeline strictly follows the description in Section 3.1 of the paper (sliding window stride of 48, label determination rules, etc.).
- You may also try retraining the model using your newly constructed CAN images to see if this yields better results.
- Our team is currently applying for authorization to release the original dataset, and we may be able to publish the complete preprocessed dataset in the future to facilitate better reproduction by the community.

### **Scenario 2: Retraining with Newly Constructed CAN Images**

If you retrained the model using your own constructed CAN images, it is crucial to pay special attention to stability issues during GAN training, particularly **mode collapse**.

During the early stages of training, if the generator (G) learns too quickly or the discriminator (D) becomes too strong, it can lead to an imbalanced adversarial training process:
- **D too strong**: The generator fails to receive effective gradient signals, causing generation quality to stagnate.
- **G too strong**: The discriminator loses its discriminative ability and cannot provide meaningful feedback.

**Recommended strategies:**
- **Learning rate decay**: In the early training phase (at least the first 200 epochs), appropriately adjust the learning rate decay strategy to ensure that G and D training progress remains relatively balanced. Monitor their loss curves to prevent one side from converging prematurely.
- **Improved loss functions**: GAN training is notoriously unstable, and the community has proposed various methods to improve training stability, such as:
  - **WGAN / WGAN-GP**: Uses Wasserstein distance instead of the original GAN's JS divergence, providing smoother gradients.
  - **LSGAN**: Adopts least squares loss to reduce gradient vanishing problems.
- **Optimizer selection**: Use more advanced optimizers (such as Adam or RMSprop) combined with gradient clipping techniques.
- **Training monitoring**: Regularly inspect the quality and diversity of generated samples to detect signs of mode collapse early.

However, in the current CAN anomaly detection task, properly adjusting the learning rate decay strategy is usually sufficient to ensure that the adversarial training remains balanced in the early stages and achieves the performance levels reported in the paper.

## **3. Applicability to Other Datasets**

We believe that the method proposed in our paper has good generalizability and can be applied to other intrusion detection datasets. We have previously tested it on image-based open set recognition datasets with promising results.

For the CICIDS2017 dataset, as you mentioned, directly constructing a 6×6 image from 30 features may carry relatively limited information. I suggest you consider the following approaches:

**1. Feature Engineering and Data Fusion:**
- You can extract and fuse features from multiple records in the PCAP files to construct a larger feature matrix (similar to how we use 48 consecutive CAN messages in our work).
- This approach can capture more temporal information and contextual correlations, resulting in converted images that contain richer discriminative features.

**2. Incorporate Design Ideas from Large Model Domain:**
- Consider borrowing methods from the large model domain, such as **Multi-Sequence Prediction (MSP)**, which incorporates more contextual information for anomaly detection.
- This approach can better leverage inter-sequence dependencies and improve the model's ability to recognize complex attack patterns.
- There are many mature optimization solutions available in practice, such as attention mechanisms and Transformer architectures, which can help the model better process sequential data.

Overall, the key is how to effectively transform raw features into input representations that contain sufficient discriminative information. If you have any questions, please feel free to continue the discussion!

Best regards

**Comment by evenchen6** [comment]
* Macro F1
<img width="801" height="377" alt="Image" src="https://github.com/user-attachments/assets/747f661d-0224-4852-b1fe-1b0506e10c87" />

* Unknown F1 (DoS)
<img width="882" height="401" alt="Image" src="https://github.com/user-attachments/assets/17f76b82-7d8e-4ed5-a5db-1d8df7bcd426" />

* val record log
`visualdl --logdir log --port 8080`
[vdlrecords.1759332732.log](https://github.com/user-attachments/files/22645109/vdlrecords.1759332732.log)
